crud/count.lua

```diff
@@ -0,0 +1,198 @@
+        scan_value = 'table',
+        tarantool_iter = 'number',
+        count_to_yield = '?number',
+        scan_condition_num = '?number',
+    })
+
+    opts = opts or {}
+
+    if opts.count_to_yield == 0 then
+        return {}
+    end
+
+    local space = box.space[space_name]
+    if space == nil then
+        return nil, CountError:new("Space %q doesn't exist", space_name)
+    end
+
+    local index = space.index[index_id]
+    if index == nil then
+        return nil, CountError:new("Index with ID %s doesn't exist", index_id)
+    end
+
+    local value = opts.scan_value
+
+    local filter_func, err = filters.gen_func(space, conditions, {
+        tarantool_iter = opts.tarantool_iter,
+        scan_condition_num = opts.scan_condition_num,
+    })
+    if err ~= nil then
+        return nil, CountError:new("Failed to generate tuples filter: %s", err)
+    end
+
+    local tuples_count = 0
+
+    local tuple
+    local gen = index:pairs(value, {iterator = opts.tarantool_iter})
+
+    gen.state, tuple = gen(gen.param, gen.state)
+    while true do
+        if tuple == nil then
+            break
+        end
+
+        local matched, early_exit = filter_func(tuple)
+
+        if matched then
+            tuples_count = tuples_count + 1
+
+            if opts.count_to_yield ~= nil and tuples_count % opts.count_to_yield == 0 then
+                fiber.yield()
+            end
+        elseif early_exit then
+            break
+        end
+
+        gen.state, tuple = gen(gen.param, gen.state)
+    end
+
+    return tuples_count
+end
+
+function count.init()
+    _G._crud.count_on_storage = count_on_storage
+end
+
+--- Calculates the number of tuples by conditions
+--
+-- @function call
+--
+-- @param string space_name
+--  A space name
+--
+-- @param ?table user_conditions
+--  Conditions by which tuples are counted
+--
+-- @tparam ?number opts.timeout
+--  Function call timeout
+--
+-- @tparam ?number opts.count_to_yield
+--  Number of tuples processed to yield after
+--
+-- @tparam ?boolean opts.prefer_replica
+--  Call on replica if it's possible
+--
+-- @tparam ?boolean opts.balance
+--  Use replica according to round-robin load balancing
+--
+-- @tparam ?string opts.mode
+--  vshard call mode
+--
+-- @return[1] number
+-- @treturn[2] nil
+-- @treturn[2] table Error description
+--
+function count.call(space_name, user_conditions, opts)
+    checks('string', '?table', {
+        timeout = '?number',
+        count_to_yield = '?number',
+        prefer_replica = '?boolean',
+        balance = '?boolean',
+        mode = '?string',
+    })
+
+    opts = opts or {}
+
+    if opts.count_to_yield ~= nil and opts.count_to_yield < 1 then
+        return nil, CountError:new("count_to_yield should be > 0")
+    end
+
+    -- check conditions
+    local conditions, err = compare_conditions.parse(user_conditions)
+    if err ~= nil then
+        return nil, CountError:new("Failed to parse conditions: %s", err)
+    end
```

Comment:
Please add a negative test with incorrect conditions.

Reply:
There has already been a test with invalid type in condition:

```lua
pgroup.test_not_valid_value_type = function(g)
    local conditions = {
        {'==', 'id', 'not_number'}
    }

    local result, err = g.cluster.main_server.net_box:call('crud.count',
            {'customers', conditions}
    )

    t.assert_equals(result, nil)
    t.assert_str_contains(err.err, "Supplied key type of part 0 does not match index part type: expected unsigned")
end
```

And I added another test with non existed field in condition:

```lua
pgroup.test_conditions_with_non_existed_field = function(g)
    local conditions = {
        {'==', 'non_existed_field', 'value'}
    }

    local result, err = g.cluster.main_server.net_box:call('crud.count',
            {'customers', conditions}
    )

    t.assert_equals(err, nil)
    t.assert_equals(result, 0)
end
```

Reply:
Neither `pgroup.test_not_valid_value_type` nor `pgroup.test_conditions_with_non_existed_field` have assert() that check error message for "Failed to parse conditions".

Reply:
Added:

```lua
pgroup.test_not_valid_operation = function(g)
    local conditions = {
        {{}, 'id', 5}
    }

    local result, err = g.cluster.main_server.net_box:call('crud.count',
            {'customers', conditions}
    )

    t.assert_equals(result, nil)
    t.assert_str_contains(err.err, "Failed to parse conditions")
end
```